test/cql-pytest/test_allow_filtering.py

```diff
@@ -308,6 +308,42 @@ def table4(cql, test_keyspace):
 
 # Selecting from indexed table should using only clustering key should require filtering
 # Variation of #7608 found in #8991
-@pytest.mark.xfail(reason="Select from indexed table using only clustering key should require filtering. Issue #7608")
 def test_select_indexed_cluster(cql, table4):
     check_af_mandatory(cql, table4, 'c1 = 1 AND c2 = 1', lambda row: row.c1 == 1 and row.c2 == 1)
+
+# table5 contains an indexed table with 3 clustering columns.
+# used to test need_filtering() on indexed tables.
+@pytest.fixture(scope="module")
+def table5(cql, test_keyspace):
+    table = test_keyspace + "." + unique_name()
+    cql.execute(f"CREATE TABLE {table} (p int, c1 text, c2 text, c3 int, PRIMARY KEY (p,c1,c2,c3))")
+    cql.execute(f"CREATE INDEX ON {table} (c3)")
+    cql.execute(f"INSERT INTO {table} (p, c1, c2, c3) VALUES (0, 'a', 'b', 0)")
+    cql.execute(f"INSERT INTO {table} (p, c1, c2, c3) VALUES (0, 'a', 'b', 1)")
+
+    everything = list(cql.execute(f"SELECT * FROM {table}"))
+    wait_for_index(cql, table, 'c3', everything)
+    yield (table, everything)
+    cql.execute(f"DROP TABLE {table}")
+
+# Test that new implementation of need_filtering() for indexes works correctly.
+# This implementation is a bit conservative - it might sometimes state
+# that filtering is needed when it isn't actually required, but at least it's safe.
```

Comment:
By saying that "it might sometimes state that filtering is needed", do you mean in the internal filtering_needed case, or does it also affect the requirement for ALLOW FILTERING? Does this test also pass on Cassandra, or not?

Reply:
The comments in the test below seem to suggest (I didn't try it myself) that it will fail on Cassandra. In this case, I personally think (and this is how the rest of test_allow_filtering.py), you should make this test pass on Cassandra, not on Scylla, and then mark the test xfail and add comments in the specific places where Scylla today fails - and add an issue to refer to it.
I personally think that everywhere we deviate from Cassandra's logic for requiring ALLOW FILTERING is a bug of Scylla, but even someone disagrees with that - it's still important to know where we deviate from it.

Reply:
They don't pass on cassandra.
I will try finding ones that work on cassandra and don't work on scylla.
These ones don't work on cassandra at all because `LIKE` requires an index :/

Reply:
Yes, I think that it's not a good idea to write tests which "enshrine" today's random behavior that doesn't have a good justification. I think that unless we have a good explanation of Cassandra's behavior being wrong or limited - Scylla should follow it. As you can see in test_allow_filtering.py today, we have a few tests which pass on Cassandra and fail on Scylla because Scylla behaves differently and there is no good explanation (other than some internal implementation details) on why we **want** it to behave differently - so it shouldn't.
So yes, I think your new test should pass on Cassandra (and you don't have to use "LIKE" is that's not a Cassandra feature - this test doesn't need LIKE...), and xfail on Scylla, and you can add comments exactly where it fails and point to an issue about it (we probably need a new issue, like we have https://github.com/scylladb/scylla/issues/7608, https://github.com/scylladb/scylla/issues/7964).

Reply:
I modified the tests and they can now run on Cassandra, but they fail.

There are some differences between Scylla and Cassandra, for example this query:
```cql
-- primary key (p, c1, c2, c3), index on c3
SELECT * FROM t WHERE p = 0 AND c1 = [1] AND c3 = 0;
```
Requires `ALLOW FILTERING` on Cassandra, but shouldn't require it on Scylla
Also there is some issue about it #4178

I added `scylla_only` to the test

Tests are supposed to pass once `need_filtering()` is perfect, for now they are marked `xfail`

I'm worried that there are still some corner cases that current `need_filtering()` misses. I added some code but it breaks tests. I need to look around more.

Reply:
> If I understand correctly, this indeed shouldn't require ALLOW FILTERING - so if I understand correctly, this can be considered a Cassandra bug

From what I heard Cassandra has different implementation of indexes than Scylla, there might be features that work on Scylla but by design are not supported on Cassandra. I will try learning more about how they are implemented in Cassandra.

Reply:
This is true. However I think the query `WHERE p = 0 AND c1 = 1 AND c3 = 0` where c3 is the indexed column, even on Cassandra can be answered by sending looking at the index for c3=0, and there, p=0 AND c1=1 should be O(1) work. Maybe calling it a Cassandra "bug" is too harsh. But my point is that if there is a case that we believe that the Cassandra behavior is "wrong" or "bad" in some sense, and we deliberately chose not to emulate it, we should explicitly say it, and mark it (with "cassandra_bug"), which is much better than just having this test fail on Cassandra and then 5 years from now our successors will wonder whether we deviated from Casssandra deliberately or by mistake.

Marking something "cassandra_bug" also requires passing a high bar of being convinced that this is a behavior we deliberately don't want to imitate. By default, when we can, we should always imitate Cassandra's behavior for features that exist in Cassandra - unless we have a very good reason to.